### ISSUE TITLE
systemd-repart fails with 4K sector size and deferred partitions

### ISSUE THREAD

**Issue by mcassaniti** [systemd/systemd#28225]
### systemd version the issue has been seen with

commit 6cc0fd0044f3a6247f80388b653fc1b09784805d (253)

### Used distribution

Ubuntu 22.10

### Linux kernel version used

5.19.0-23-generic

### CPU architectures issue was seen on

x86_64

### Component

systemd-repart

### Expected behaviour you didn't see

Partitions get created exactly the same regardless of sector size.

### Unexpected behaviour you saw

Padding is miscalculated when adding previously deferred partitions.

### Steps to reproduce the problem

Use the following configuration in the `repart.d` directory and run the commands as shown in the program output.

```shell
# cat 00-ESP.conf
[Partition]
Type         = esp
Label        = ESP
Format       = vfat

SizeMinBytes = 128M
SizeMaxBytes = 128M

# Sufficient for testing
CopyFiles    = /etc:/
```

```shell
# cat 10-os.conf
[Partition]
Type           = root
Label          = test
Format         = squashfs

Minimize       = best
# Sufficient for testing
CopyFiles      = /etc/:/

VerityMatchKey = os
Verity         = data
```

```shell
# cat 11-os-verity.conf
[Partition]
Type           = root-verity
Label          = test

Minimize       = best

VerityMatchKey = os
Verity         = hash
```

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
root@test-machine:/root/artifacts/image# # Set sector size for VFAT to 512 bytes because there will not be enough FAT clusters otherwise
root@test-machine:/root/artifacts/image# export SYSTEMD_REPART_MKFS_OPTIONS_VFAT=-S 512
root@test-machine:/root/artifacts/image# sector_size=4096
root@test-machine:/root/artifacts/image# systemd-repart --dry-run=no --empty=create --size=auto --sector-size=$sector_size --defer-partitions esp --no-pager --definitions /root/src/repart.d --no-pager image.raw
Pre-populating squashfs filesystem of partition 10-os.conf twice to calculate minimal partition size
Populating squashfs filesystem.
Successfully populated squashfs filesystem.
Parallel mksquashfs: Using 8 processors
Creating 4.0 filesystem on /var/tmp/.#repartcd0ad5f39c93f783, block size 131072.
[==================================================================================================================================================================================================|] 459/459 100%

Exportable Squashfs 4.0 filesystem, gzip compressed, data block size 131072
        compressed data, compressed metadata, compressed fragments,
        compressed xattrs, compressed ids
        duplicates are removed
Filesystem size 359.68 Kbytes (0.35 Mbytes)
        35.19% of uncompressed filesystem size (1022.16 Kbytes)
Inode table size 13247 bytes (12.94 Kbytes)
        26.46% of uncompressed inode table size (50067 bytes)
Directory table size 12731 bytes (12.43 Kbytes)
        50.76% of uncompressed directory table size (25079 bytes)
Number of duplicate files found 16
Number of inodes 1086
Number of files 467
Number of fragments 7
Number of symbolic links 478
Number of device nodes 0
Number of fifo nodes 0
Number of socket nodes 0
Number of directories 141
Number of hard-links 0
Number of ids (unique uids + gids) 4
Number of uids 3
        root (0)
        unknown (65436)
        nobody (65534)
Number of gids 4
        root (0)
        unknown (65436)
        shadow (42)
        nogroup (65534)
/var/tmp/.#repartcd0ad5f39c93f783 successfully formatted as squashfs (no label or uuid specified)
Minimal partition size of squashfs filesystem of partition 10-os.conf is 360.0K
Pre-populating verity hash data of partition 11-os-verity.conf to calculate minimal partition size
Minimal partition size of verity hash partition 11-os-verity.conf is 8.0K
Automatically determined minimal disk image size as 149.0M.
Sized 'image.raw' to 149.0M.
Applying changes.
TYPE               LABEL UUID                                 FILE              NODE             SIZE PADDING ROOTHASH
esp                ESP   27ae6c41-e269-4b66-9920-d604b581f752 00-ESP.conf       image.raw1   → 128.0M    → 0B 
root-x86-64        test  345e5091-d414-02b7-7f07-dff7328de843 10-os.conf        image.raw2    → 10.0M    → 0B 345e5091d41402b77f07dff7328de84365e02bc4084d707a0b0018ee18e98817
root-x86-64-verity test  65e02bc4-084d-707a-0b00-18ee18e98817 11-os-verity.conf image.raw3    → 10.0M    → 0B 345e5091d41402b77f07dff7328de84365e02bc4084d707a0b0018ee18e98817
                                                                                           Σ = 148.0M  Σ = 0B 

 ▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░
 │                                                                  │    └─ 11-os-verity.conf
 │                                                                  └─ 10-os.conf
 └─ 00-ESP.conf                                                                

Wiped block device.
Discarded entire block device.
Successfully wiped file system signatures from future partition 1.
Successfully wiped file system signatures from future partition 2.
Copying in '/var/tmp/.#repartcd0ad5f39c93f783' (360.0K) on block level into future partition 1.
Copying in of '/var/tmp/.#repartcd0ad5f39c93f783' on block level completed.
Copying in '/var/tmp/.#repart2d1a37aca6baf61e' (8.0K) on block level into future partition 2.
Copying in of '/var/tmp/.#repart2d1a37aca6baf61e' on block level completed.
Adding new partition 1 to partition table.
Adding new partition 2 to partition table.
Writing new partition table.
All done.
TYPE               LABEL UUID                                 FILE              NODE             SIZE PADDING ROOTHASH
esp                ESP   27ae6c41-e269-4b66-9920-d604b581f752 00-ESP.conf       image.raw1   → 128.0M    → 0B 
root-x86-64        test  345e5091-d414-02b7-7f07-dff7328de843 10-os.conf        image.raw2    → 10.0M    → 0B 345e5091d41402b77f07dff7328de84365e02bc4084d707a0b0018ee18e98817
root-x86-64-verity test  65e02bc4-084d-707a-0b00-18ee18e98817 11-os-verity.conf image.raw3    → 10.0M    → 0B 345e5091d41402b77f07dff7328de84365e02bc4084d707a0b0018ee18e98817
                                                                                           Σ = 148.0M  Σ = 0B

root@test-machine:/root/artifacts/image# systemd-repart --dry-run=no --empty=allow --size=auto --sector-size=$sector_size --no-pager --definitions /root/src/repart.d --no-pager image.raw                        
Automatically determined minimal disk image size as 149.0M, current image size is 149.0M.
File '/root/artifacts/image/image.raw' already is of requested size or larger, not growing. (149.0M >= 149.0M)
Applying changes.
TYPE               LABEL UUID                                 FILE              NODE                                      SIZE   PADDING ROOTHASH
esp                ESP   27ae6c41-e269-4b66-9920-d604b581f752 00-ESP.conf       /root/artifacts/image/image.raw1      → 128.0M      → 0B 
root-x86-64        test  345e5091-d414-02b7-7f07-dff7328de843 10-os.conf        /root/artifacts/image/image.raw2 10.0M → 10.0M 4.0K → 0B TBD
root-x86-64-verity test  65e02bc4-084d-707a-0b00-18ee18e98817 11-os-verity.conf /root/artifacts/image/image.raw3 10.0M → 10.0M 4.0K → 0B TBD
                                                                                                                    Σ = 148.0M    Σ = 0B 

 ▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░
 │                                                                  │    └─ 11-os-verity.conf
 │                                                                  └─ 10-os.conf
 └─ 00-ESP.conf                                                                

Successfully wiped file system signatures from future partition 0.
Successfully discarded data from future partition 0.
Formatting future partition 0.
Populating vfat filesystem.
Successfully populated vfat filesystem.
mkfs.fat 4.2 (2021-01-31)
skipping directory symlink user
/var/tmp/repart-0KiRcb successfully formatted as vfat (label "ESP", uuid 8e85a9d5)
Successfully formatted future partition 0.
Adding new partition 0 to partition table.
Growing existing partition 1.
Failed to update partition: Numerical result out of range
```

**Comment by yuwata** [comment]
Fix is waiting in #28228.

**Comment by mcassaniti** [comment]
@yuwata I can't thank you enough for the quick turn around.
